Add "Excluded Paths" for Game Directories settings in general settings glade

avikpl1911 opened the pull request:
Issue: https://github.com/Ryujinx/Ryujinx/issues/5911

Included excluded directories and added excluded directories .Hence, making changes to file format for the config

![image](https://github.com/Ryujinx/Ryujinx/assets/65647597/19a547bd-4e5f-4f82-b273-68440aee8cfe)

piplup55 commented:
Settings fail to open on linux
[Ryujinx_1.0.0-dirty_2024-02-08_06-06-38.log](https://github.com/Ryujinx/Ryujinx/files/14204932/Ryujinx_1.0.0-dirty%2B42be9fa5884b28723a4fc925daa0ab8841552b68_2024-02-08_06-06-38.log)

avikpl1911 commented:
Can you show me your config.json

avikpl1911 commented:
@piplup55

piplup55 commented:
> Can you show me your config.json

![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/3f804718-7bb8-423b-b5c6-5398f548014c)

avikpl1911 commented:
In place of null put []

avikpl1911 commented:
I don't know why that is i will fix it

avikpl1911 commented:
In exclude_game_dirs

avikpl1911 commented:
@piplup55

piplup55 commented:
> I don't know why that is i will fix it

it's likely because ava ui was opened first before gtk, as this change is only for gtk ava ui might be getting confused and setting it to null

avikpl1911 commented:
Ok I get it now

avikpl1911 commented:
I've to make fixes on Ava

piplup55 commented:
not 100% sure this is a bug but the exclude list doesn't work on gtk if the directory is in the game list
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/aeab92a4-8717-4b43-81a9-3453dd30bee8)
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/7188da6e-59d9-4a0a-8710-d7916869f018)
same for ava ui
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/fe05f43b-cc4e-4f0e-a3eb-27270f099b2f)

avikpl1911 commented:
This is not actually a bug but it can be fixed it's so that exclude games dir is only for storing the strings but can be fixed

avikpl1911 commented:
ive added changes to file config version and also fixed the null error in config.json and added exclusion features to the applicaionloader library it also supports legacy ava @piplup55 you can check it out now all features are working @TSRBerry implementation of the feature is complete you can test it out and pls see if i have done the naming conventions right as in docs

avikpl1911 commented:
@marysaka 

@AcK77 

@emmauss

@TSRBerry

can you review this pr

avikpl1911 commented:
is there a problem with this pr?

piplup55 commented:
most issues are still preset plus extras since the latest commit, game now are randomly removed from the list if excluded
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/74ddab3b-6d7b-48e4-ae00-a8657631ad22)
GTK now crashes on boot if it's null
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/5aa6b5a5-6860-49cf-99b5-d105354378f6)
Extra Debugging lines?
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/3c46dbf7-3beb-4024-866b-f95e6eb903f3)
ava ui still not having any options to exclude
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/ca15ddd1-53c5-4a58-a896-82c51d7621ab)
ava ui ignoring the setting completely
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/3e113b8d-2c13-44e1-a5cd-fb323dbd102e)

avikpl1911 commented:
@piplup55 games are only randomly removed if the dir inside the given dir is excluded

avikpl1911 commented:
If u implicitly put the option to null gtk will crash the change depends upon the config files version which modifies the value if not present
As directed by any other change to the config.json instance Ava ui I've not worked on because I didn't get any response on this pr

piplup55 commented:
> @piplup55 games are only randomly removed if the dir inside the given dir is excluded

if you look at my excluded paths, you will see that my whole nsp folder is excluded so it shouldn't load any nsp titles into the list yet it still does

avikpl1911 commented:
Does it happen for avalonia or both

piplup55 commented:
> Does it happen for avalonia or both

the resetting to null only happens if ava is opened

avikpl1911 commented:
And wtb the dir included prob

WilliamWsyHK commented on src/Ryujinx/Ui/Windows/SettingsWindow.cs:
```diff
@@ -770,18 +837,80 @@ private void AddDir_Pressed(object sender, EventArgs args)
             ((ToggleButton)sender).SetStateFlags(StateFlags.Normal, true);
         }
 
+        private void Include_exclude_Pressed(object sender,EventArgs args)
```
`Include_Exclude_Pressed`

WilliamWsyHK commented on src/Ryujinx/Ui/Windows/SettingsWindow.cs:
```diff
@@ -723,11 +770,31 @@ private void SystemTimeSpin_ValueChanged(object sender, EventArgs e)
         }
 
+
+        private void onRowActivated(object sender, ButtonReleaseEventArgs args)
```
Update the method name? I don't have a good name at the moment, but it should be `PascalCase` like `OnRowActivated`.

WilliamWsyHK commented on src/Ryujinx/Ui/Windows/SettingsWindow.cs:
```diff
@@ -723,11 +770,31 @@ private void SystemTimeSpin_ValueChanged(object sender, EventArgs e)
+            _gameDirsBoxExcluded.Selection.UnselectAll();
+        }
+
+        private void onRowActivated1(object sender, ButtonReleaseEventArgs args)
```
Definitely not `OnRowActivated1`, without stating the intention of the function, and just add 1 after the method name after the previous function.

WilliamWsyHK commented on src/Ryujinx/Ui/Windows/SettingsWindow.cs:
```diff
@@ -770,18 +837,80 @@ private void AddDir_Pressed(object sender, EventArgs args)
+                    _gameDirsBoxStoreIncluded.Remove(ref treeIter);
+                }
+
+            ((ToggleButton)sender).SetStateFlags(StateFlags.Normal, true);
```
Should indent 1 time.

WilliamWsyHK commented on src/Ryujinx/Ui/Windows/SettingsWindow.cs:
```diff
@@ -770,18 +837,80 @@ private void AddDir_Pressed(object sender, EventArgs args)
+                    _gameDirsBoxStoreExcluded.Remove(ref treeIter1);
+                }
+
+            ((ToggleButton)sender).SetStateFlags(StateFlags.Normal, true);
```
Should indent 1 time.

WilliamWsyHK commented on src/Ryujinx/Ui/Windows/SettingsWindow.cs:
```diff
@@ -770,18 +837,80 @@ private void AddDir_Pressed(object sender, EventArgs args)
 
             ((ToggleButton)sender).SetStateFlags(StateFlags.Normal, true);
+            }
+
```
Extra empty lines.

avikpl1911 commented on src/Ryujinx/Ui/Windows/SettingsWindow.cs:
```diff
@@ -37,6 +41,12 @@ public class SettingsWindow : Window
         private float _previousVolumeLevel;
         private bool _directoryChanged = false;
 
+        private enum IncludeorExclude { Include, Exclude }
```
Yes will do
